### PR TITLE
Karl Design Doc.

### PR DESCRIPTION
**Pull request by dkorolev**
@mzhurovich : PTAL, thank you!

### CODE REVIEW

File: Karl/README.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+**Optional: { Alert Tags }.**
+
+#### Dispatcher: The logic that generates pages given alerts and mutes
+
+A single instance per Karl.
+
+By default, muted alerts do not become pages, and non-muted ones are promoted to pages and sent to all DRIs. Paged alerts are then automatically muted by their { Service, Name }, any tags, for the next 15 minutes.
+
+The logic to take into consideration current muting configuration, alert severity, DRI configuration, account medias (e-mail, SMS, etc.), time of day, vacations, etc., is implemented as the Dispatcher.
```
Comment (sompylasar) [inline Karl/README.md]: Could not parse the sentence.

File: Karl/README.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+Alerts are defined and configured in a C++ file. No bare strings are used; each entity is defined as a variable of certain type, and then passed over.
+
+A running Karl exposes a dashboard. Karl is also Claire: it is safe to run under `cron`, and it is friendly with continuous integration.
+
+Everything is stored in Yoda, with a REST-ful browsing API for viewing current configuration, scrapes, their results, alerts, pages, and mutes. Mutes can be set, prolonged/shrinked, and unset via this API.
+
+`TODO(dkorolev): Discuss aggregation master-plan with Max.`
+
+For Claire binaries requiring extra time to start up or tear down, pass `karl::WarmupPeriod(FLAGS_warmup_period_ms)` and/or `karl::ShutdownPeriod(FLAGS_shutdown_period_ms)`. This will tell Claire to allow for more time before switching form graceful to non-graceful shutdown of an already running binary in case it is by design expected to take more than the default three seconds to start up or to tear down.
```
Comment (sompylasar) [inline Karl/README.md]: `form graceful` -> `from graceful`

Comment (sompylasar) [inline Karl/README.md]: `in case it is by design expected to take` -> `in case it is expected by design to take`

File: Karl/README.md (hunk 1 above)
Comment (dkorolev) [inline Karl/README.md]: Yes, it was by far not the best way to phrase it.

PTAL @ d46754b

File: Karl/README.md (hunk 2 above)
Comment (dkorolev) [inline Karl/README.md]: Both fixed in d46754b.

File: Karl/README.md (hunk 1 above)
Comment (sompylasar) [inline Karl/README.md]: Way better, thanks!

File: Karl/README.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
+# Karl
+
+*Your service should be up and running. Up and running, Karl!*
+
+Karl is Current's monitoring, alerting, and continous integration service.
```
Comment (mzhurovich) [inline Karl/README.md]: s/continous/continuous/

File: Karl/README.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+// * `state_uptime_ms`:   The time the binary has been running in this state, in millis.
+// * `local_time`:        To ensure NTP is up and running locally.
+// * `binary_build_time`: To track how up to date the presently running version is.
+CLAIRE(FLAGS_port);
+```
+
+Also, Claire enables continuous integration with no extra work.
+
+```cpp
+// Checks whether another Claire-powered binary is already running on this port.
```
Comment (mzhurovich) [inline Karl/README.md]: Smth like `On startup ...` in the beginning of the sentence is missing.

File: Karl/README.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+**A Directly Responsible Individual who may receive alerts.**
+
+Has a unique name. For the purposes of this doc, assume they have an e-mail address on file.
+  
+  
+#### Service
+
+**A collection of servers to monitor.**
+
+Has a unique name. Each Server to belongs to one and only Service.
```
Comment (mzhurovich) [inline Karl/README.md]: s/to belongs/belongs/

File: Karl/README.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+## Example
+
+### Client: Claire
+
+Claire is a binary monitored by Karl.
+
+```cpp
+// Registers a `/current` HTTP endpoint on the specified port. It responds with a JSON containing:
+// * `uptime_ms`:         The uptime of the binary.
+// * `state`:             "StartingUp", "Running", or "TearingDown" <=> HTTP 500, 200 and 404 codes.
```
Comment (mzhurovich) [inline Karl/README.md]: 500 -> 503?

File: Karl/README.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+  
+   Servers also have names, and each name should be unique within the service they belong to.
+
+   For alerting, paging and muting purposes, each server belongs to one and only one service.
+
+```cpp
+DRI(dima, EMail("dmitry.korolev@gmail.com"));
+SERVICE(backend);
+SERVER(api, backend, "http://api.current.ai");
+
```
Comment (mzhurovich) [inline Karl/README.md]: DRIs should be somehow assigned to Services, right?

File: Karl/README.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+CLAIRE_WITH_CONTINUOUS_INTEGRATION(FLAGS_port);
+```
+
+The above design makes continuous integration as straightforward as inserting:
+* `(cd $DIR; ./$BINARY`), or even
+* `(cd $DIR; git pull; make $BINARY; ./$BINARY)  # Yay for Linux-way!`
+
+into `crontab`.
+
+To monitor extra fields, simple, trivially copyable types can be declared as `std::atomic_*`-s and registered as `CLAIRE_VAR()`-s.
```
Comment (mzhurovich) [inline Karl/README.md]: How they are integrated into the response structure described above?

File: Karl/README.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+Claire is a binary monitored by Karl.
+
+```cpp
+// Registers a `/current` HTTP endpoint on the specified port. It responds with a JSON containing:
+// * `uptime_ms`:         The uptime of the binary.
+// * `state`:             "StartingUp", "Running", or "TearingDown" <=> HTTP 500, 200 and 404 codes.
+// * `state_uptime_ms`:   The time the binary has been running in this state, in millis.
+// * `local_time`:        To ensure NTP is up and running locally.
+// * `binary_build_time`: To track how up to date the presently running version is.
```
Comment (mzhurovich) [inline Karl/README.md]: Add command line arguments?

File: Karl/README.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+//
+// * If the same binary of the same build time is already running, it will keep running,
+//   and the binary that was starting on top of it will silently exit with success exit code.
+//
+// * If the version already running is older than the one that is just being started,
+//   the currently starting binary will gracefully tear down the older one and start itself.
+//
+// * In any other case, or in case of failure, an alert is fired.
+//
+// Exposes `/current?sigterm` and `/current?stop` to enable the above.
```
Comment (mzhurovich) [inline Karl/README.md]: As I understand, these two correspond to some sort of graceful/non-graceful shutdown?
What if I have the situation when the binary just stopped answering over HTTP for some reason?

File: Karl/README.md (hunk 7 above)
Comment (dkorolev) [inline Karl/README.md]: Yes, I though "everything pages everyone" is a solid default :-)

File: Karl/README.md (hunk 9 above)
Comment (dkorolev) [inline Karl/README.md]: Thought of it a lot. Dropped for two reasons: 1) extra code to write, 2) a potential security hole.

File: Karl/README.md (hunk 6 above)
Comment (dkorolev) [inline Karl/README.md]: Done.

File: Karl/README.md (hunk 3 above)
Comment (dkorolev) [inline Karl/README.md]: Done.

File: Karl/README.md (hunk 4 above)
Comment (dkorolev) [inline Karl/README.md]: :+1:

File: Karl/README.md (hunk 10 above)
Comment (dkorolev) [inline Karl/README.md]: Great question! Let's leave it for a phone chat .. some of the next ones, not the one today.

File: Karl/README.md (hunk 8 above)
Comment (dkorolev) [inline Karl/README.md]: Those are a `map<string, string>`. More complex ones their own response types, controlled by the URL parameter. The easiest I could think of for now.

File: Karl/README.md (hunk 5 above)
Comment (dkorolev) [inline Karl/README.md]: :+1:

**Pull request merged**